File: packages/ui/db/entity/Parking.ts
import BaseColumnSchemaPart from "@/db/entity/BaseColumnSchemaPart";
import { type IParking } from "models";
import { EntitySchema } from "typeorm";
import CoordinateSchema, { Coordinate } from "./Coordinate";
import { Crag } from "./Crag";

export type Parking = Omit<IParking, "crag" | "location"> & {
  crag?: Crag;
  cragId: number | null;
  location: Coordinate;
};

const ParkingSchema = new EntitySchema<Parking>({
  name: "parking",
  columns: {
    ...BaseColumnSchemaPart,
    name: {
      type: String,
      nullable: true,
    },
    description: {
      type: String,
      nullable: true,
    },
    cragId: {
      type: "int",
      nullable: true,
    },
  },
  relations: {
    crag: {
      type: "one-to-one",
      target: "crag",
      joinColumn: true,
      onDelete: "SET NULL",
      cascade: ["insert", "update"],
    },
  },
  embeddeds: {
    location: {
      schema: CoordinateSchema,
      prefix: "location_",
    },
  },
});

export default ParkingSchema;

File: go-server/models/parking.go
package models

type Parking struct {
	ID          *int       `json:"id,omitzero"`
	Name        *string    `json:"name,omitzero"`
	Description *string    `json:"description,omitzero"`
	Location    Coordinate `json:"location"`
	Crag        *Crag      `json:"crag,omitzero"`
}

func (p Parking) IsZero() bool {
	return p.ID == nil && p.Name == nil && p.Description == nil && p.Location.IsZero() && p.Crag == nil
}
